Jerrod/Assignments/Rails_models/dojo_ninjas.txt

```diff
@@ -0,0 +1,118 @@
+Dojo.first.ninjas.create(first_name: 'John', last_name:'Johnson')
+	create three ninjas that belong to the first Dojo
+
+FIFTH***************************************
+Dojo.second.ninjas.create(first_name:'Sam', last_name:'Winchester')
+Dojo.second.ninjas.create(first_name:'Dean', last_name:'Winchester')
+Dojo.second.ninjas.create(first_name:'Adom', last_name:'Winchester')
+	create three new ninjas that belong to the second dojo
+SIXTH***************************************
+Dojo.third.ninjas.create(first_name:'Michael', last_name:'ArchAngel')
+Dojo.third.ninjas.create(first_name:'Lucifer', last_name:'ArchAngel')
+Dojo.third.ninjas.create(first_name:'Gabriel', last_name:'ArchAngel')
+	create three ninjas for the third dojo
+SEVENTH***************************************
+Dojo.first.ninjas.all
++----+------------+-----------+---------+-------------------------+-------------------------+
+| id | first_name | last_name | dojo_id | created_at              | updated_at              |
++----+------------+-----------+---------+-------------------------+-------------------------+
+| 1  | first      | last      | 4       | 2016-12-02 03:08:59 UTC | 2016-12-02 03:08:59 UTC |
+| 2  | Louis      | Tom       | 4       | 2016-12-02 03:09:16 UTC | 2016-12-02 03:09:16 UTC |
+| 3  | John       | Johnson   | 4       | 2016-12-02 03:09:26 UTC | 2016-12-02 03:09:26 UTC |
++----+------------+-----------+---------+-------------------------+-------------------------+
+Dojo.second.ninjas.all
+Dojo.third.ninjas.all
+	Know how to get all of the ninjas for any dojo
+
+EIGHT***************************************
+Dojo.second.ninjas.select(:first_name).order("created_at DESC")
+	retrieve the first_names of the ninjas that belong to the second dojo and order the result by created_at DESC order
+
+NINTH***************************************
+	Delete the second dojo. How could you adjust your model so that it 			automatically removes all of the ninjas associated with that dojo?
+	**************FIRST WAY I DID THIS**********
+Dojo.second.ninjas.destroy_all
+Dojo.second.destroy
+	Delete the second dojo. How could you adjust your model so that it 			automatically removes all of the ninjas associated with that dojo?
+
+	One way is shown above, first destroy all the ninjas associated with it. 		Delete_all only deletes the reference, so you have to do destroy_all to 		delete the actual ninja
+
+
+	*************SECOND WAY I DID IT************
+class Dojo < ActiveRecord::Base
+  has_many :ninjas
+  validates :state, length: {is: 2}
+  validates :name, :city, :state, presence: true
+
+
+  before_destroy :delete_ninjas
+
+  def delete_ninjas
+    Dojo.find(self.id).ninjas.destroy_all
+  end
+end
+
+Rails Console Commands******
+Dojo.find(6).destroy
+	
+	
+	Second method, I created a before_destroy method in the dojo.rb model, so that it would destroy_all on any ninja with that Dojo id as it’s reference, tested it twice, worked, it destroyed all ninjas associated with it then itself.
+
```

Comment:
Making your own 'before_destroy :delete_ninjas' method is a cool way to do it, but there's a much easier way.  On your Dojo model all you need to do is add one thing to the has_many line:

     has_many :ninjas, dependent: :destroy

With dependent: :destroy all you need to do is 'Dojo.find(6).destroy' and it will automatically delete all of the associated ninjas without having to create your own method.

Here's the other difference between using destroy/destroy_all vs delete/delete_all.  For dependent: :destroy to work (and also for your before_destroy as well), you have to use destroy/destroy_all.  Delete/delete_all ignores those and would only delete the Dojos, leaving the ninjas intact in the database, including the dojo_id, which obviously wouldn't exist anymore.